chainstate/src/detail/best_chain_candidates.rs

```diff
@@ -0,0 +1,177 @@
+// Copyright (c) 2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::BTreeSet;
+
+use thiserror::Error;
+
+use super::chainstateref::ChainstateRef;
+use crate::TransactionVerificationStrategy;
+use chainstate_storage::BlockchainStorageRead;
+use chainstate_types::{BlockIndex, GenBlockIndex, PropertyQueryError};
+use common::{chain::Block, primitives::Id, Uint256};
+use utils::tap_error_log::LogError;
+
+#[derive(Eq, Copy, Clone, Debug)]
+pub struct BestChainCandidatesItem {
+    pub chain_trust: Uint256,
+    pub block_id: Id<Block>,
+}
+
+impl BestChainCandidatesItem {
+    fn from_block_index(block_index: &BlockIndex) -> Self {
+        BestChainCandidatesItem {
+            block_id: *block_index.block_id(),
+            chain_trust: block_index.chain_trust(),
+        }
+    }
+}
+
+impl Ord for BestChainCandidatesItem {
```

Comment:
Initially, the struct had an extra field that didn't participate in the ordering, so I had to have a manual implementation. But after removing the field I decided to keep the manual implementation, because without it the correctness of the comparison would depend on the order of the fields in the struct, which seemed a bit fragile.

Reply:
> the comparison would depend on the order of the fields

But do we care unless it's consistent. The problem is that it's easy to forget to update the method if a field is added. I'd go with derives for Ord and also for PartialOrd and PartialEq below.

Reply:
The items must be sorted by chain_trust first, that's the whole point of the set.
Also, it's a kind of struct where when adding a field you should first ask yourself whether it should participate in the ordering in the first place.
Anyway, I don't think that this struct will ever be changed (other than being removed completely), so I don't mind getting rid of the manual impl in favour of a comment that would state that chain_trust should always come first.

Reply:
[discussed] I've written a comment near Ord::cmp mentioning that the order of fields is important and that chain_trust must come first. Otherwise the code stays the same.